Cygwin/X: GLX crash workaround

Remove GL dispatcher symbols which are defined in libGL, and link 
with libGL for them instead (as dri_swrast.so is linked with libGL)

Paper over the cracks. Avoid a crash when we have a broken context.

Turn on GLX null context debugging

---
 configure.ac         |   10 +++++-----
 glx/Makefile.am      |    2 --
 glx/single2.c        |    3 +++
 hw/xwin/InitOutput.c |   24 ++++++++++++++++++++++++
 4 files changed, 32 insertions(+), 7 deletions(-)

Index: xorg-server-1.5.3/xserver/configure.ac
===================================================================
--- xorg-server-1.5.3.orig/xserver/configure.ac	2008-11-20 21:38:21.506369300 -0600
+++ xorg-server-1.5.3/xserver/configure.ac	2008-11-23 18:42:42.934276800 -0600
@@ -829,7 +829,7 @@
 	AC_SUBST(XLIB_CFLAGS)
 	AC_DEFINE(GLXEXT, 1, [Build GLX extension])
 	GLX_LIBS='$(top_builddir)/glx/libglx.la'
-	GLX_SYS_LIBS="$GLX_SYS_LIBS $DLOPEN_LIBS"
+	GLX_SYS_LIBS="$GLX_SYS_LIBS $GL_LIBS $DLOPEN_LIBS"
 else
         GLX=no
 fi
@@ -1680,7 +1680,7 @@
 			XWIN_SYS_LIBS=-lwinsock2
 			;;
 	esac
-	XWIN_SYS_LIBS="$XWIN_SYS_LIBS $XWINMODULES_LIBS"
+	XWIN_SYS_LIBS="$XWIN_SYS_LIBS $XWINMODULES_LIBS $GLX_SYS_LIBS"
 	AC_SUBST(XWIN_SERVER_NAME)
 	AC_SUBST(XWIN_SYS_LIBS)
 
Index: xorg-server-1.5.3/xserver/glx/Makefile.am
===================================================================
--- xorg-server-1.5.3.orig/xserver/glx/Makefile.am	2008-11-05 10:52:17.000000000 -0600
+++ xorg-server-1.5.3/xserver/glx/Makefile.am	2008-11-23 18:51:08.559276800 -0600
@@ -38,10 +38,8 @@
 	dispatch.h				\
 	glapitable.h				\
 	glapitemp.h				\
-	glapi.c					\
 	glapi.h					\
 	glapioffsets.h				\
 	glprocs.h				\
-	glthread.c				\
 	glthread.h
 
 libglxdri_la_SOURCES = \  
Index: xorg-server-1.5.3/xserver/glx/single2.c
===================================================================
--- xorg-server-1.5.3.orig/xserver/glx/single2.c
+++ xorg-server-1.5.3/xserver/glx/single2.c
@@ -341,6 +341,9 @@ int DoGetString(__GLXclientState *cl, GL
     string = (const char *) CALL_GetString( GET_DISPATCH(), (name) );
     client = cl->client;
 
+    if (string == NULL)
+      string = "";
+
     /*
     ** Restrict extensions to those that are supported by both the
     ** implementation and the connection.  That is, return the
Index: xorg-server-1.5.3/xserver/hw/xwin/InitOutput.c
===================================================================
--- xorg-server-1.5.3.orig/xserver/hw/xwin/InitOutput.c
+++ xorg-server-1.5.3/xserver/hw/xwin/InitOutput.c
@@ -134,6 +134,9 @@ const char *
 winGetBaseDir(void);
 #endif
 
+static
+void glx_debugging(void);
+
 /*
  * For the depth 24 pixmap we default to 32 bits per pixel, but
  * we change this pixmap format later if we detect that the display
@@ -1045,6 +1048,8 @@ InitOutput (ScreenInfo *screenInfo, int 
        * Apply locale specified in LANG environment variable.
        */
       setlocale (LC_ALL, "");
+
+      glx_debugging();
     }
 #endif
 
@@ -1127,3 +1132,22 @@ winCheckDisplayNumber ()
 
   return TRUE;
 }
+
+/* GLX debugging helpers */
+#include <../glx/glapi.h>
+
+static
+void warn_func(void * p1, const char *format, ...) {
+  va_list v;
+  va_start(v, format);
+  vfprintf(stderr, format, v);
+  va_end(v);
+  fprintf(stderr,"\n");
+}
+
+static
+void glx_debugging(void)
+{
+  _glapi_set_warning_func(warn_func);
+  _glapi_noop_enable_warnings(TRUE);
+}
